Add LoadDiffCal support for alternate grouping in HDF format.

bolotovskyr opened the pull request:
**Description of work**

`LoadDiffCal` algorithm allows to specify an alternate detector grouping through the `GroupFilename` property. However, at the moment, only ".cal" and ".xml" formats are supported.   At the same time, `SaveDiffCal` can produce a grouping definition in the HDF format. 

This PR extends the `GroupFilename` functionality by allowing these new file extensions: ".h5", ".hd5", ".hdf".  The existing `GroupFilename` unit test (with XML-formatted grouping definition) is now extended to include a check of the resultant grouping workspace. Also, a new unit test is added for an HDF-formatted grouping definition.  

**To test**

`../mantid/build/bin/DataHandlingTest LoadDiffCalTest`


#### Reviewer ####

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

#### Gatekeeper ####

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

AnthonyLim23 committed:
refs #35442 added quickBayes lorentzians

AnthonyLim23 committed:
refs 35442 adding quickbayes

AnthonyLim23 committed:
refs 35442 adding quickbayes

AnthonyLim23 committed:
refs #35442 add bayesStretch2

AnthonyLim23 committed:
refs #35442 add bayesStretch2

peterfpeterson committed:
Change pystog to be a conda dependency

rbauststfc committed:
Add unit tests for matching preview row on title and theta

rbauststfc committed:
Add method for parsing runtitle strings

rbauststfc committed:
Find lookup rows on preview tab by workspace title and angle

rbauststfc committed:
Use title parsing method in SearchResult.cpp

thomashampson committed:
Fix boost::python related compilation error.

thomashampson committed:
Fix deprecated and removed functions

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

Commit:
Fix archive on macOS

RE #35802

Commit:
Add a release note

RE #35802

Commit:
Amend test for macOS

RE #35802

rbauststfc committed:
Display run title on preview tab

rbauststfc committed:
Check if loaded workspace exists in LookupTable::findLookupRow

rbauststfc committed:
Fix preview tab crash if reduction runs with no loaded ws

rbauststfc committed:
Add release notes for issue 35704 final fix

rbauststfc committed:
Improve a couple of type declarations

AnthonyLim23 committed:
refs #35442 add first draft of bayesstretch2

thomashampson committed:
Temporarily change environment name while developing

thomashampson committed:
Up python to 3.10

Needed to remove the setuptools pin to resolve dependencies.

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

thomashampson committed:
Fix collections.Iterables removal

Iterables has been moved to collections.abc and was derecated from within collections
a while ago. In Python 3.10 it was removed.

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

thomashampson committed:
Update collections imports for python 3.10

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

thomashampson committed:
Explicitly convert to ints in QPoint and QSize

Python 3.10 no longer uses implicit conversion with loss from a float to an int

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

jhaigh0 committed:
add workspaces to outwards group so algorithm history finds them

thomashampson committed:
Add new op codes introduced in python 3.9

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>
Co-authored-by: Martyn Gigg <martyn.gigg@stfc.ac.uk>

thomashampson committed:
Fix QPoint implicit conversion errors

thomashampson committed:
Fix failing test

jhaigh0 committed:
added test to check that history has been coppied

jhaigh0 committed:
call fillHistory in child workspaces to pass on the algorithm history

jhaigh0 committed:
added release note]

thomashampson committed:
Add test for lhs_info() function in funcinspect.py

jhaigh0 committed:
add try except and better feedback to the user

jhaigh0 committed:
added a release note

AnthonyLim23 committed:
refs #35442 bayes code clean up

jclarkeSTFC committed:
Use variables for the items in the two combo boxes, and add them in code, instead of in the .ui file.

jclarkeSTFC committed:
Rearranged layout of TOF converter window.

Made the window have centred boxes, in a fixed layout, instead of manually setting box sizes and location. Deleted some redundant text labels.

jclarkeSTFC committed:
Deleted try-except logic in imports.

This try-except was added in 2018, I think it was related to some refactoring. Doesn't seem required now unless I've missed something.

jclarkeSTFC committed:
Disable things if units are the same.

If the input and output units are the same, then disable several things in the interface.

AnthonyLim23 committed:
refs #35442 adding tests for new bayes algs

jclarkeSTFC committed:
Don't clear TOF inputs unnecessarily.

Fixes #35585. The disabling of the boxes at the start of this method is not needed, instead it only disables them if not applicable to the current units. This also tidied up the logic further down the method.

thomashampson committed:
Add more tests for lhs_info function.

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

thomashampson committed:
Fix issue with unpacking dictionaries inside function calls and using lhs_info

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

jhaigh0 committed:
tidy up of funcinspect

jclarkeSTFC committed:
Normalise input angle to [0, 360)

The scattering angle should be in [0, 360), otherwise when converting to d-spacing you end up dividing by sin(angle > pi) which is going to give you a negative d-spacing.

jclarkeSTFC committed:
Disable angle for d-spacing to momentum transfer

When converting between d-spacing and momentum transfer the scattering angle is not required, although it is required when either of those units is used with another unit.

AnthonyLim23 committed:
refs #35442 adding tests and docs for new bayes algs

jclarkeSTFC committed:
Add release notes.

AnthonyLim23 committed:
refs #35442 adding tests for new bayes algs

AnthonyLim23 committed:
refs #35442 adding tests for new bayes algs

thomashampson committed:
Fix problem with GIL release in runningInstancesOf

Limited the scope of the GIL release so that manipulating python objects
can be done within the function. Previously it worked, but the behaviour was
changed for Python 3.10. See here:
https://docs.python.org/3/c-api/dict.html#c.PyDict_GetItem

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>
Co-authored-by: Martyn Gigg <martyn.gigg@stfc.ac.uk>

thomashampson committed:
Update conda recipe to Python 3.10

Commit:
update rebin with new param

rebin changes and new tests

some documentation changes, release notes, two small changes

minor edits for cleaner code

fix failing doc test, add unit test to prevent in future

fix typo, make test pass

Changes to align with refactor of EnumeratedString

fix for simpleapi docs test

non-cosmetic changes removed from exec

binmode not needed in exec

more reversions to rebin

add test not covered to default params behavior

fidgeting updates to rebin tests

improvements to validation and tests

updates to fix rebin tests

AnthonyLim23 committed:
refs #35442 adding tests for new bayes algs

AnthonyLim23 committed:
refs #35442 adding tests for new bayes algs

robertapplin committed:
Import dos dependency from inside functions
The dos dependency is stored in the Mantid scripts directory. This
directory cannot be imported at the moment of subscribing the
algorithm to the AlgorithmFactory. As a consequence, when building
the docs, the SimulatedDensityOfStates algorithm had a missing
documentation file.

thomashampson committed:
Make ConfigService base directory compatible with long paths on Windows

WorkbenchStartupTest was failing in Python 3.10. Something was prepending
the Windows long path syntax to the MANTIDPATH env variable. You cannot mix
back and forward slashes in long paths, so here we have addressed that.

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

thomashampson committed:
Remove quasielasticbayes install from Windows standalone package

It's not compatible with Python 3.10. We will replace it with quickBayes
when that is ready.

robertapplin committed:
Build StandardTestData for mantiddocs package

robertapplin committed:
Fix missing file due to interpreting file name as a run

robertapplin committed:
Configure the datasearch directories before building docs-qthelp

robertapplin committed:
Treat sphinx warnings as errors in docs packaging

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

AnthonyLim23 committed:
refs #35442 adding tests for new bayes algs

AnthonyLim23 committed:
Merge branch 'main' into 35442_quickBayes

thomashampson committed:
Reset name of development environment

thomashampson committed:
Added release note

robertapplin committed:
Explicitly import simpleapi algorithms to fix sphinx warning

robertapplin committed:
Run pre-commit black on tube_calib.py

robertapplin committed:
Remove references to mtd

AnthonyLim23 committed:
refs #35442 adding tests for new bayes algs

jclarkeSTFC committed:
Update TOF Converter docs.

thomashampson committed:
Update cppcheck suppression

AnthonyLim23 committed:
refs #35442 adding tests for new bayes algs

AnthonyLim23 committed:
refs #35442 adding tests for new bayes algs

thomashampson committed:
fix bullet in release notes

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate

updates:
- [github.com/astral-sh/ruff-pre-commit: v0.0.282 → v0.0.284](https://github.com/astral-sh/ruff-pre-commit/compare/v0.0.282...v0.0.284)

jclarkeSTFC committed:
Add axes labels.

Axes labels on plot inside the Sample Transmission Calculator window.

jclarkeSTFC committed:
Error check on bin width.

Don't allow the bin width to be greater than the upper bin limit.

jclarkeSTFC committed:
Improve multiple bin info.

Add a tooltip to the edit box for entering the bins when using the Multiple option. When there aren't an odd number of values in the box then explain why this is a problem in the error message.

jclarkeSTFC committed:
Make errors more visible.

Now it draws a red box around the error message.

jclarkeSTFC committed:
Add release notes

jhaigh0 committed:
add validator to check workspace has multiple bins

AnthonyLim23 committed:
refs #35442 fix failing tests

jhaigh0 committed:
add suggested changes

jhaigh0 committed:
Remove deprecated PyEval_InitThreads()

This function will do nothing in Python>=3.9 and was been called from Py_Initialize() since 3.7
https://docs.python.org/3/c-api/init.html#c.PyEval_InitThreads

jhaigh0 committed:
add test to test_figurewindow

jhaigh0 committed:
added release note

jhaigh0 committed:
Update Framework/Algorithms/src/UnGroupWorkspace.cpp

Co-authored-by: Rob Applin <40830825+robertapplin@users.noreply.github.com>

jclarkeSTFC committed:
Add single-value workspaces as input option.

Add any single-value workspaces as options to a new combo box, so you can either pick one of those, or enter a specified value. It will remember the last specified value that you typed in when you switch input method.

jclarkeSTFC committed:
Update test cases.

The validation messages have changed, and the width is now checked relative to the max, so the tests needed updating. I also added a new test for checking the width > max validation message.

bolotovskyr committed:
Add support for alternate grouping in HDF format.

bolotovskyr committed:
Add support for alternate grouping in HDF format.

rboston628 committed:
Merge branch 'main' into ewm1851_updateRebin

bolotovskyr committed:
Refactor file extension check.

bolotovskyr committed:
Refactor file extension check.

bolotovskyr committed:
Consolidate filename extension checks. Add release notes.

bolotovskyr committed:
Consolidate filename extension checks. Add release notes.

jmborr committed:
Add property OutputScalingWorkspace

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
add release notes

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
property is enough to be a string

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

AnthonyLim23 committed:
refs #35442 clean up output for bayes code

AnthonyLim23 committed:
refs #35442 clean up output for bayes code

jhaigh0 committed:
Update docs/source/release/v6.8.0/Workbench/Bugfixes/34397.rst

Co-authored-by: Sarah Foxley <55837273+sf1919@users.noreply.github.com>

jclarkeSTFC committed:
Update release notes.

Comment about using single valued workspaces was missing.

jclarkeSTFC committed:
Move constants to top of file.

jclarkeSTFC committed:
Fix some logic and reduce duplication.

Fixed logic for determining whether it's a momentum transfer to d-spacing conversion or not, and put the logic in one place only.

jclarkeSTFC committed:
Add comment.

bolotovskyr committed:
Investigating Windows compiler complaint...

bolotovskyr committed:
Investigating Windows compiler complaint...

AnthonyLim23 committed:
refs #35442 clean up output for bayes code

AnthonyLim23 committed:
refs #35442 clean up output for bayes code

bolotovskyr committed:
Investigating Windows compiler complaint contd...

bolotovskyr committed:
Investigating Windows compiler complaint contd...

bolotovskyr committed:
Investigating Windows compiler complaint contd....

bolotovskyr committed:
Investigating Windows compiler complaint contd....

bolotovskyr committed:
Use EnumeratedString.

bolotovskyr committed:
Use EnumeratedString.

bolotovskyr committed:
Merge branch 'main' into ewm1775_save_grouping_file

sf1919 committed:
Remove importlib_resources call in workbench meta.yaml

When the pin for importlib_resources was removed the call to the pin in the meta.yaml file for workbench was missed. This meant that on the nightly build it failed because it couldn't find the pin that was being referenced.

jclarkeSTFC committed:
Update release notes

Co-authored-by: Jonathan Haigh <35813666+jhaigh0@users.noreply.github.com>

jclarkeSTFC committed:
Give figure a tight layout

Prevents axes running off the edge of the plot area.

Co-authored-by: Jonathan Haigh <35813666+jhaigh0@users.noreply.github.com>

jclarkeSTFC committed:
Add info from tooltip to docs.

AnthonyLim23 committed:
Merge branch 'main' into 35442_quickBayes

AnthonyLim23 committed:
refs #35442 clean up output for bayes code

jclarkeSTFC committed:
Improve ui resize behaviour.

Previously if you changed the window width then the new space would appear between the box labels and the input boxes. I've added spacers so the labels stay next to the input boxes.

thomashampson committed:
Merge pull request #35906 from mantidproject/0-fix-doc-warnings

Nightly failure when building `docs-qthelp`

bolotovskyr committed:
Further usage of EnumeratedString.

bolotovskyr committed:
Further usage of EnumeratedString.

bolotovskyr committed:
Merge branch 'ewm1775_save_grouping_file' of https://github.com/mantidproject/mantid into ewm1775_save_grouping_file

rbauststfc committed:
Reset row state when settings changed via preview tab

rbauststfc committed:
Add release notes for issue 35565

peterfpeterson committed:
Merge pull request #35936 from mantidproject/stitch1d_scalingws

Extend Mantid's Stitch1D to return SingleValueWorkspace with scaling + scaling_error

thomashampson committed:
Merge pull request #35940 from mantidproject/tidy_workbench_meta

Remove importlib_resources call in workbench meta.yaml

AnthonyLim23 committed:
refs #35442 clean up output for bayes code

AnthonyLim23 committed:
Apply suggestions from code review

robertapplin committed:
Fix ruff isinstance warnings in mantid Framework and qt

robertapplin committed:
Fix ruff isinstance warnings in scripts directory

robertapplin committed:
Remove xvfb wrap temporarily

Set QT_QPA_PLATFORM env var to offscreen

robertapplin committed:
Fix latex warnings

rboston628 committed:
improvements to tests of rebin

rboston628 commented:
Code is of high quality.  Extra test case added for the added input option.  Release notes added.  Needed changes to documentation will be handled by the property descriptions.

jhaigh0 committed:
suggested variable rename

rbauststfc committed:
Remove incorrect steps after polarization correction

Skip sum banks and background subtraction steps when
ReflectometryReductionOneAuto is called again after polarization
correction

rbauststfc committed:
Add release notes for issue 35855

AnthonyLim23 committed:
refs #35442 clean up output for bayes code

AnthonyLim23 committed:
refs #35442 fix failing test

AnthonyLim23 committed:
refs #35442 fix failing test

Commit:
Replace C-style strings with string_view

RE #35802

gemmaguest committed:
Merge pull request #35870 from mantidproject/34387_handle_bad_plot_script_generation

Handle error thrown by the plot script generator when generating an unsupported plot.

rboston628 committed:
add test requested for param validator

gemmaguest committed:
Merge pull request #35826 from mantidproject/35704_fix_lookup_matching_from_preview_tab

Fix lookup matching from ISIS Reflectometry Preview tab

gemmaguest committed:
Merge pull request #35886 from mantidproject/33898_tof_converter_improvements

Improvements and bug fixes for the TOF Converter window

gemmaguest committed:
Merge pull request #35930 from mantidproject/35011_sample_transmission_calculator_improvements

Fix bugs in the Sample Transmission Calculator

gemmaguest committed:
Merge pull request #35865 from mantidproject/34255_ungroupworkspace_history

Add missing algorithm history tracking to UnGroupWorkspace

gemmaguest committed:
Merge pull request #35932 from mantidproject/32474_drag_overplot_validator

Avoid bug when attempting to overplot single bin workspace

thomashampson committed:
Merge pull request #35944 from robertapplin/0-fix-doc-qthelp-warnings

Better output when building `docs-qthelp` and fix warnings

robertapplin commented on docs/source/release/v6.8.0/Framework/Data_Handling/New_features/35935.rst:
```diff
@@ -0,0 +1 @@
+* :ref:`LoadDiffCal <algm-LoadDiffCal>` now accepts an alternate grouping definition in the HDF format.
```
Can you use a `-` instead of `*` for the bullet point? Using a `*` can cause issues further down the line for the Release Editors
```suggestion
- :ref:`LoadDiffCal <algm-LoadDiffCal>` now accepts an alternate grouping definition in the HDF format.
```

Commit:
Fix plot not appearing when running centre finder

Seems like workbench.app.mainwindow is being imported later in the
workbench startup than it was before, Using workbench.app seems to work
just as well.

RE #35931

peterfpeterson committed:
Convert pystog to an optional dependency

bolotovskyr committed:
Update docs/source/release/v6.8.0/Framework/Data_Handling/New_features/35935.rst

Co-authored-by: Rob Applin <40830825+robertapplin@users.noreply.github.com>

bolotovskyr committed:
Update docs/source/release/v6.8.0/Framework/Data_Handling/New_features/35935.rst

Co-authored-by: Rob Applin <40830825+robertapplin@users.noreply.github.com>

bolotovskyr commented:
@robertapplin, I have committed your suggestion. However, it turned our this PR needs a bit more work, so please don't approve/merge it. Thanks!

bolotovskyr committed:
EnumeratedString with templated string comparator.

bolotovskyr committed:
EnumeratedString with templated string comparator.

robertapplin committed:
Update ruff to 0.285

MialLewis committed:
update path to include md5 on m1 macs

robertapplin committed:
Apply more isinstance changes

AndreiSavici committed:
Merge pull request #35896 from mantidproject/ewm1851_updateRebin

Refactor of Rebin

robertapplin committed:
Fix last isinstance warning

Commit:
Avoid constructing unneeded string object

RE #35802

bolotovskyr committed:
Further implementation and tests of case-insensitive EnumeratedString.

bolotovskyr committed:
Further implementation and tests of case-insensitive EnumeratedString.

bolotovskyr committed:
Bugfix and updated comments.

robertapplin committed:
Build docs-doctest target without using xvfb for better output

SilkeSchomann committed:
Merge pull request #35923 from mantidproject/pre-commit-ci-update-config

[pre-commit.ci] pre-commit autoupdate

SilkeSchomann committed:
Merge pull request #35803 from ConorMFinn/35802_macos_archive

Allow the Archive To Be Accessed From macOS

SilkeSchomann committed:
Merge pull request #35801 from mantidproject/35442_quickBayes

35442 quick bayes

SilkeSchomann committed:
Merge pull request #35915 from mantidproject/35228_python_3_10_upgrade

Upgrade to Python 3.10

SilkeSchomann committed:
Merge pull request #35955 from mantidproject/fix_mamba_install_on_m1_mac

Update path to include md5 on M1 macs during mambaforge install

SilkeSchomann committed:
Merge pull request #35673 from peterfpeterson/pystog_conda

Change pystog to be a conda dependency

robertapplin committed:
Set QT_QPA_PLATFORM on same line as cmake call to limit its scope

jhaigh0 committed:
Rename Memory widget to System Memory Usage Widget

jhaigh0 committed:
Include instructions for Mac for finding the about page

robertapplin committed:
Archive doctest log

jhaigh0 committed:
Remove specific roles for each test to simplify

jhaigh0 committed:
Add manual test for plot rename via workspace from #35089

robertapplin committed:
Archive doctest logs in same folder as ctest logs

jhaigh0 committed:
remove lingering red text role

gemmaguest committed:
Merge pull request #35942 from mantidproject/35565_reset_row_state_from_preview_apply

Reset row state when experiment settings are changed via the Preview tab

bolotovskyr committed:
Implement customizable string comparison.

bolotovskyr committed:
Using std::function, still need to update the docs.

bolotovskyr committed:
Update the doc.

gemmaguest committed:
Merge pull request #35965 from mantidproject/35021_update_core_workbench_testing

Update Core manual testing instructions

SilkeSchomann committed:
Merge pull request #35952 from ConorMFinn/35951_beam_centre_finder_plot

Fix plot not appearing when running centre finder

bolotovskyr committed:
Move definition of firstLetterComparator closer to where it's used.

gemmaguest committed:
Merge pull request #35945 from mantidproject/35855_fix_background_sub_with_polarization_analysis

Remove incorrect steps after polarization correction in ReflectometryReductionOneAuto

SilkeSchomann committed:
Merge pull request #35964 from robertapplin/0-better-output-for-doctests-on-nightly

Output doctest logs on nightly

peterfpeterson committed:
Merge pull request #35970 from mantidproject/ewm2236_enum_string_case_insensitive

Support for customizable string comparison in EnumeratedString

bolotovskyr committed:
Merge branch 'ewm1775_save_grouping_file' of https://github.com/mantidproject/mantid into ewm1775_save_grouping_file

bolotovskyr committed:
Adapt LoadDiffCal code to the latest EnumeratedString implementation.
